src/time_zone_name_win.cc

```diff
@@ -0,0 +1,177 @@
+      return ucal_getTimeZoneIDForWindowsIDRef;
+    }
+  }
+
+  const std::wstring system32_dir = GetSystem32Dir();
+  if (system32_dir.empty()) {
+    g_unavailable.store(true, std::memory_order_relaxed);
+    return nullptr;
+  }
+
+  // Here LoadLibraryExW(L"icu.dll", nullptr, LOAD_LIBRARY_SEARCH_SYSTEM32) does
+  // not work if "icu.dll" is already loaded from somewhere other than the
+  // system32 directory. Specifying the full path with LoadLibraryW is more
+  // reliable.
+  const std::wstring icu_dll_path = system32_dir + L"\\icu.dll";
+  const HMODULE icu_dll = ::LoadLibraryW(icu_dll_path.c_str());
+  if (icu_dll == nullptr) {
+    g_unavailable.store(true, std::memory_order_relaxed);
+    return nullptr;
+  }
+
+  const auto ucal_getTimeZoneIDForWindowsIDRef =
+      AsProcAddress<ucal_getTimeZoneIDForWindowsID_func>(
+          icu_dll, "ucal_getTimeZoneIDForWindowsID");
+  if (!ucal_getTimeZoneIDForWindowsIDRef) {
+    g_unavailable.store(true, std::memory_order_relaxed);
+    return nullptr;
+  }
+
+  g_ucal_getTimeZoneIDForWindowsID.store(ucal_getTimeZoneIDForWindowsIDRef,
+                                         std::memory_order_relaxed);
+
+  return ucal_getTimeZoneIDForWindowsIDRef;
+}
+
+// Convert wchar_t array (UTF-16) to UTF-8 string
+std::string Utf16ToUtf8(const wchar_t* ptr, size_t size) {
+  if (size > std::numeric_limits<int>::max()) {
+    return std::string();
+  }
+  const int chars_len = static_cast<int>(size);
+  std::string result;
+  // TODO: Remove std::max() when we require C++17 or higher.
+  auto len = std::max<size_t>(result.capacity(), 1);
+  do {
+    result.resize(len);
+    // TODO: Switch to std::string::data() when we require C++17 or higher.
+    len = ::WideCharToMultiByte(CP_UTF8, WC_ERR_INVALID_CHARS, ptr, chars_len,
+                                &result[0], static_cast<int>(result.size()),
+                                nullptr, nullptr);
+  } while (len > result.size());
+  result.resize(len);
+  return result;
+}
+
+}  // namespace
+
+std::string GetWindowsLocalTimeZone() {
+  const auto getTimeZoneIDForWindowsID = LoadIcuGetTimeZoneIDForWindowsID();
+  if (getTimeZoneIDForWindowsID == nullptr) {
+    return std::string();
+  }
+
+  DYNAMIC_TIME_ZONE_INFORMATION info = {};
+  if (::GetDynamicTimeZoneInformation(&info) == TIME_ZONE_ID_INVALID) {
+    return std::string();
+  }
+
+  std::wstring result;
+  // TODO: Remove std::max() when we require C++17 or higher.
+  auto len = std::max<size_t>(result.capacity(), 1);
+  while (true) {
+    UErrorCode status = U_ZERO_ERROR;
+    result.resize(len);
+    // TODO: Switch to std::string::data() when we require C++17 or higher.
+    len = getTimeZoneIDForWindowsID(
+        info.TimeZoneKeyName, -1, nullptr, &result[0],
+        static_cast<std::int32_t>(result.size()), &status);
+    if (U_SUCCESS(status)) {
+      return Utf16ToUtf8(result.data(), len);
+    }
+    if (status != U_BUFFER_OVERFLOW_ERROR) {
```

Comment:
I would probably also fail if `len != static_cast<std::int32_t>(len)`, as that would cause us to pass the wrong length on the next loop.

Indeed, that test should come before the `getTimeZoneIDForWindowsID()` call in case `result.capacity()` is initially beyond `std::int32_t` limits.  That's highly unlikely, I know, but why not get it exactly right anyway.

Alternatively, the initial `len` could be capped at the maximum `std::int32_t`.

Reply:
Done.